server/src/compilerDiagnostics/diagnostics/SpecifyCompilerVersion.ts

```diff
@@ -0,0 +1,78 @@
+import { TextDocument } from "vscode-languageserver-textdocument";
+import {
+  CodeAction,
+  CodeActionKind,
+  Diagnostic,
+  Range,
+} from "vscode-languageserver/node";
+import { CompilerDiagnostic, ResolveActionsContext } from "../types";
+import { attemptConstrainToFunctionName } from "../conversions/attemptConstrainToFunctionName";
+import { HardhatCompilerError, ServerState } from "../../types";
+
+/**
+ * This diagnostic is shown when no compiler version is specified
+ * i.e. pragma solidity XXXX;
+ *
+ * The suggested quickfix adds the pragma solidity statement, by
+ * extracting the version from hardhat's warning message.
+ */
+export class SpecifyCompilerVersion implements CompilerDiagnostic {
+  public code = "3420";
+  public blocks: string[] = [];
+
+  public fromHardhatCompilerError(
+    document: TextDocument,
+    error: HardhatCompilerError
+  ): Diagnostic {
+    return attemptConstrainToFunctionName(document, error);
+  }
+
+  public resolveActions(
+    _serverState: ServerState,
+    _diagnostic: Diagnostic,
+    { uri, document }: ResolveActionsContext
+  ): CodeAction[] {
+    // Get the compiler specification code from hardhat warning
+    const regex = /pragma solidity .*;/;
+    const match = _diagnostic.message.match(regex);
+    const pragmaLine = match && match[0];
+
+    if (pragmaLine === null) {
+      return [];
+    }
+
+    // If diagnostic is shown on "// SPDX ..." line, insert on next line.
+    // Otherwise insert on previous line
+    const position = { character: 0, line: _diagnostic.range.end.line };
+
+    const checkRange: Range = {
+      start: _diagnostic.range.start,
+      end: {
+        line: _diagnostic.range.end.line,
+        character: _diagnostic.range.end.character + 1,
+      },
+    };
+
+    if (document.getText(checkRange) === "/") {
```

Comment:
I think this covers the most common case, but I am tempted to say we tighten this up to do a check for whether it is a valid `SPDX-License-Identifier` line. If it is just a comment I think the rule should be that we put the compiler pragma on the first line and ignore the comment, only if the first line is a `SPDX-License-Identifier` line does the compiler pragma move from first to second line.